### PR TITLE
Allow the removal of an in use template if there are other matching ones

### LINKED ISSUE #96022: Highest priority data stream index template can't be deleted
### Elasticsearch Version

8.9.0-SNAPSHOT built from b5af53d

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

cloud container

### Problem Description

When an index template is considered to be the highest priority template for an existing data stream it can't be delete, even if a matching index template with a lower priority exists. That means it's not possible to revert a data stream to a lower-priority template without applying a work-around. But as long as there is a matching index template to fall back to, this should be allowed. This is a common sequence of steps when installing and later uninstalling an integration package.

### Steps to Reproduce

Create a logs data stream based on the built-in `logs` index template by indexing a document:

```
POST logs-felixstuermer.test-default/_doc
{
  "@timestamp": "2023-05-05T12:00:00"
}
```

Observe that the data stream considers the default template as its matching template:

```
GET _data_stream/logs-felixstuermer.test-default
```

```json
{
  "data_streams": [
    {
      "name": "logs-felixstuermer.test-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-logs-felixstuermer.test-default-2023.05.11-000001",
          "index_uuid": "moqOcaWvSzG6tMBhCpitjg"
        }
      ],
      "generation": 1,
      "_meta": {
        "description": "default logs template installed by x-pack",
        "managed": true
      },
      "status": "GREEN",
      "template": "logs", // <--
      "ilm_policy": "logs",
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false
    }
  ]
}
```

Create a higher-priority index template:

```
PUT _index_template/logs-felixstuermer.test-0
{
  "index_patterns": ["logs-felixstuermer.test-default"],
  "data_stream": { },
  "composed_of": [],
  "priority": 500
}
```

Observe that the data stream now considers the default template as its matching template, even if no rollover has happened:

```
GET _data_stream/logs-felixstuermer.test-default
```

```json
{
  "data_streams": [
    {
      "name": "logs-felixstuermer.test-default",
      "timestamp_field": {
        "name": "@timestamp"
      },
      "indices": [
        {
          "index_name": ".ds-logs-felixstuermer.test-default-2023.05.11-000001",
          "index_uuid": "moqOcaWvSzG6tMBhCpitjg"
        }
      ],
      "generation": 1,
      "_meta": {
        "managed": true,
        "description": "default logs template installed by x-pack"
      },
      "status": "GREEN",
      "template": "logs-felixstuermer.test-0", // <--
      "hidden": false,
      "system": false,
      "allow_custom_routing": false,
      "replicated": false
    }
  ]
}
```

Try to delete the high-priority index template and observe that this fails:

```
DELETE _index_template/logs-felixstuermer.test-0
```

```json
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "unable to remove composable templates [logs-felixstuermer.test-0] as they are in use by a data streams [logs-felixstuermer.test-default]"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "unable to remove composable templates [logs-felixstuermer.test-0] as they are in use by a data streams [logs-felixstuermer.test-default]"
  },
  "status": 400
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by gmarouli**
A data stream can match multiple composable index templates which are prioritised and the highest priority template is the currently in-use template. A data stream should always have a matching template, for this reason when we are removing a composable template we check that it is not in use. If it is in use, we throw an error and the removal fails.

However, since a data stream can match multiple templates, removing the currently in use template could mean that the data stream would start using the next inline. In this case, the removal should be allowed. 

This is what this PR addresses. When we try to remove a composable template, we do not check if this template is in use, but if this template is the only template that a data stream could use. Only then we fail the removal, otherwise the removal succeeds and the data stream starts using the next in-line template.

Fixes: #96022

### CODE REVIEW

**Review by dakrone — APPROVED** [review]
This LGTM, thanks Mary! I left a few really minor comments about nullability and documentation.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 1]
```diff
@@ -1183,54 +1194,71 @@ public static List<IndexTemplateMetadata> findV1Templates(Metadata metadata, Str
...
+
+        return winnerName;
+    }
+
+    /**
+     * Return an ordered list of the name (id) and composable index templates that would apply to an index. The first
+     * one is the winner template that is applied to this index. In the event that no templates are matched,
+     * {@code null} is returned.
+     */
+    @Nullable
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1228]: You are right! Fixed.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 2]
```diff
@@ -966,7 +966,7 @@ public static ClusterState innerRemoveIndexTemplateV2(ClusterState currentState,
         return ClusterState.builder(currentState).metadata(metadata).build();
     }
 
-    static Set<String> dataStreamsUsingTemplates(final ClusterState state, final Set<String> templateNames) {
+    static Set<String> dataStreamsExclusivelyUsingTemplates(final ClusterState state, final Set<String> templateNames) {
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:969]: Yes, I will.

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 3]
```diff
@@ -1183,54 +1194,71 @@ public static List<IndexTemplateMetadata> findV1Templates(Metadata metadata, Str
...
-                    + String.join(",", winner.composedOf())
-                    + "] defined the index.hidden setting, which is not allowed"
-            );
-        }
+        CollectionUtil.timSort(candidates, Comparator.comparing(candidate -> candidate.v2().priorityOrZero(), Comparator.reverseOrder()));
+        return candidates;
+    }
 
-        return winnerName;
+    private static boolean hasIndexHiddenSetting(Metadata metadata, ComposableIndexTemplate winner, String winnerName) {
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:1259]: Missed that! Yes, I will fix it.

**Comment by gmarouli** [comment]
@elasticmachine update branch

**Comment by ruflin** [comment]
Great to see this landed!

**Pull request merged**